core/services/llo/offchain_config_digester.go

```diff
@@ -0,0 +1,126 @@
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/llo-feeds/generated/exposed_channel_verifier"
+)
+
+// Originally sourced from: https://github.com/smartcontractkit/offchain-reporting/blob/991ebe1462fd56826a1ddfb34287d542acb2baee/lib/offchainreporting2/chains/evmutil/offchain_config_digester.go
+
+var _ ocrtypes.OffchainConfigDigester = OffchainConfigDigester{}
+
+func NewOffchainConfigDigester(chainID *big.Int, contractAddress common.Address) OffchainConfigDigester {
+	return OffchainConfigDigester{chainID, contractAddress}
+}
+
+type OffchainConfigDigester struct {
+	ChainID         *big.Int
+	ContractAddress common.Address
+}
+
+func (d OffchainConfigDigester) ConfigDigest(cc ocrtypes.ContractConfig) (ocrtypes.ConfigDigest, error) {
+	signers := []common.Address{}
+	for i, signer := range cc.Signers {
+		if len(signer) != 20 {
+			return ocrtypes.ConfigDigest{}, errors.Errorf("%v-th evm signer should be a 20 byte address, but got %x", i, signer)
+		}
+		a := common.BytesToAddress(signer)
+		signers = append(signers, a)
+	}
+	transmitters := []credentials.StaticSizedPublicKey{}
+	for i, transmitter := range cc.Transmitters {
+		if len(transmitter) != 2*ed25519.PublicKeySize {
+			return ocrtypes.ConfigDigest{}, errors.Errorf("%v-th evm transmitter should be a 64 character hex-encoded ed25519 public key, but got '%v' (%d chars)", i, transmitter, len(transmitter))
+		}
+		var t credentials.StaticSizedPublicKey
+		b, err := hex.DecodeString(string(transmitter))
+		if err != nil {
+			return ocrtypes.ConfigDigest{}, errors.Wrapf(err, "%v-th evm transmitter is not valid hex, got: %q", i, transmitter)
+		}
+		copy(t[:], b)
+
+		transmitters = append(transmitters, t)
+	}
+
+	return configDigest(
+		d.ChainID,
+		d.ContractAddress,
+		cc.ConfigCount,
+		signers,
+		transmitters,
+		cc.F,
+		cc.OnchainConfig,
+		cc.OffchainConfigVersion,
+		cc.OffchainConfig,
+	), nil
+}
+
+func (d OffchainConfigDigester) ConfigDigestPrefix() (ocrtypes.ConfigDigestPrefix, error) {
+	return ocrtypes.ConfigDigestPrefixLLO, nil
+}
+
+func makeConfigDigestArgs() abi.Arguments {
+	abi, err := abi.JSON(strings.NewReader(exposed_channel_verifier.ExposedChannelVerifierABI))
+	if err != nil {
+		// assertion
+		panic(fmt.Sprintf("could not parse aggregator ABI: %s", err.Error()))
+	}
+	return abi.Methods["exposedConfigDigestFromConfigData"].Inputs
+}
+
+var configDigestArgs = makeConfigDigestArgs()
+
+func configDigest(
+	chainID *big.Int,
+	contractAddress common.Address,
+	configCount uint64,
+	oracles []common.Address,
+	transmitters []credentials.StaticSizedPublicKey,
+	f uint8,
+	onchainConfig []byte,
+	offchainConfigVersion uint64,
+	offchainConfig []byte,
+) types.ConfigDigest {
+	msg, err := configDigestArgs.Pack(
+		chainID,
+		contractAddress,
+		configCount,
+		oracles,
+		transmitters,
+		f,
+		onchainConfig,
+		offchainConfigVersion,
+		offchainConfig,
+	)
+	if err != nil {
+		// assertion
+		panic(err)
```

Comment:
why? i don't like seeing panics in application logic while those applications are running in embedded in the core node. it assumes that either 1. the panic will be recovery, which is out of scope of the app or 2. it's ok for the node to crash, which is also out of scope for the app

Reply:
Disagree. Panics are good when we want to assert something that should _never_ happen, and if they do happen, something has gone so wrong we can't trust the application to continue running.

Reply:
@jmank88 probably has an opinion on this

Reply:
I think I already commented on this case. It doesn't make sense to me to panic when the caller has the ability to return an error, and in general it is always more flexible to make that decision higher up the stack.

Also, even if this should never happen, it doesn't necessarily mean that we can't trust the application. We may have just hit an edge case in the ABI packing. It is not as if we encountered some broken state invariant that would affect future calls.

Reply:
FWIW the cases below seem slightly different, but they could be improved to at least panic with a message that includes the relevant information.